### PR TITLE
Percent decode automatically in `extract::Path`

### PR DESCRIPTION
**Pull request by davidpdrsn**
Fixes https://github.com/tokio-rs/axum/issues/261

@jplatte what you think?

### CODE REVIEW

**Comment by lnicola** [comment]
Just throwing out some random thoughts:

 - not every framework does percent decoding, it might be less surprising to have a special extractor for it
 - if not, this should be clearly mentioned in the `Path` documentation
 - a `PathBuf` extractor would also useful, and should probably percent-decode automatically (and check for path traversal too)
 - in my app I also want non-UTF-8 paths to work (using https://crates.io/crates/os-str-bytes), but that's more of a niche use case

**Comment by davidpdrsn** [comment]
> not every framework does percent decoding, it might be less surprising to have a special extractor for it

I think it should be the other way around. Have `Path` decode by default, and then provide a more niche extractor (possibly in another crate) that doesn't. I imagine working with utf8 is more common in rust.

> if not, this should be clearly mentioned in the Path documentation

Definitely!

> a PathBuf extractor would also useful, and should probably percent-decode automatically (and check for path traversal too)

That's orthogonal to this but is something I've thought about as well. Contributions are welcome 😊

**Comment by lnicola** [comment]
Yeah, the raw path should still be available in the request, so that's fine.

File: src/util/mod.rs [hunk 1]
```diff
@@ -34,6 +34,39 @@ impl ByteStr {
...
+    pub(crate) fn new<S>(s: S) -> Self
+    where
+        S: AsRef<str>,
+    {
+        let byte_str = if let Ok(decoded) =
+            percent_encoding::percent_decode(s.as_ref().as_bytes()).decode_utf8()
+        {
+            ByteStr::new(decoded)
+        } else {
+            ByteStr::new(s)
```
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply src/util/mod.rs]: I think I managed to find a fix. The extractor will now fail if one of the params contained invalid utf-8 after decoding.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,13 +7,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 # Unreleased
 
-- None.
+- **fixed:** Fix `extract::Path` not percent decoding params ([#272](https://github.com/tokio-rs/axum/pull/272))
```
Comment (jplatte) [inline CHANGELOG.md:10]: This should be `**changed**` and the PR labeled 'breaking change'. Doing more or less percent encoding clearly changes observable behavior.

File: src/extract/path/de.rs [hunk 3]
```diff
@@ -561,18 +560,16 @@ mod tests {
...
     {
-        UrlParams(
-            values
-                .into_iter()
-                .map(|(k, v)| (ByteStr::new(k), ByteStr::new(v)))
-                .collect(),
-        )
+        values
+            .into_iter()
+            .map(|(k, v)| (ByteStr::new(k), PercentDecodedByteStr::new(v).unwrap()))
```
Comment (jplatte) [inline src/extract/path/de.rs:571]: I think this `.unwrap()` should be an `.expect("message")`. Without having read the rest of this PR I can't tell why this wouldn't sometimes cause a panic.

File: src/util/mod.rs [hunk 4]
```diff
@@ -34,6 +34,39 @@ impl ByteStr {
...
+    {
+        let byte_str = if let Ok(decoded) =
+            percent_encoding::percent_decode(s.as_ref().as_bytes()).decode_utf8()
+        {
+            ByteStr::new(decoded)
+        } else {
+            return None;
+        };
+
+        Some(Self(byte_str))
```
Comment (jplatte) [inline src/util/mod.rs:53]: This seems unnecessarily verbose with the "early" `return`, but I guess that's a just matter of taste.

File: src/routing/mod.rs [hunk 5]
```diff
@@ -626,24 +626,49 @@ where
     }
 }
 
-#[derive(Debug)]
-pub(crate) struct UrlParams(pub(crate) Vec<(ByteStr, ByteStr)>);
+// we store the potential error here such that users can handle invalid path
+// params using `Result<Path<T>, _>`. That wouldn't be possible if we
+// returned an error immediately when decoding the param
+pub(crate) struct UrlParams(
+    pub(crate) Result<Vec<(ByteStr, PercentDecodedByteStr)>, InvalidUtf8InPathParam>,
```
Comment (jplatte) [inline src/routing/mod.rs:633]: I wonder whether this should be `Vec<Result<(ByteStr, PercentDecodedByteStr), InvalidUtf8InPathParam>>`. Would make sense even without really "supporting" non-UTF8 segments if the path deserializer supported [`IgnoredAny`](https://docs.rs/serde/1.0.130/serde/de/struct.IgnoredAny.html).

File: CHANGELOG.md (hunk 2 above)
Comment (davidpdrsn) [reply CHANGELOG.md:10]: Bah you're right. One more thing that has to wait for 0.3 😞

File: src/extract/path/de.rs (hunk 3 above)
Comment (davidpdrsn) [reply src/extract/path/de.rs:571]: This is test code.

File: src/routing/mod.rs (hunk 5 above)
Comment (davidpdrsn) [reply src/routing/mod.rs:633]: I think I'll just leave it as it. Can always improve things if someone has a use case.

File: src/util/mod.rs (hunk 4 above)
Comment (davidpdrsn) [reply src/util/mod.rs:53]: Sure

**Pull request merged**